### ISSUE TITLE
Baseline file is ignored when `--diff` option is used

### ISSUE THREAD

**Issue by Wirone** [vimeo/psalm#2191]
Given that I have a project with configured Psalm and error baseline path, when I run analysis I don't see any errors:

```
-> % dc exec php vendor/bin/psalm --show-info=false                  
Scanning files...
Analyzing files...

░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  60 / 304 (20%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 120 / 304 (39%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 180 / 304 (59%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 240 / 304 (79%)
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 300 / 304 (99%)
░░░░

------------------------------
No errors found!
------------------------------

Checks took 22.50 seconds and used 551.933MB of memory
Psalm was able to infer types for 90.9964% of the codebase
```

However when I change one file and run analysis with `--diff` I see more errors than newly introduced ones:

```
-> % dc exec php vendor/bin/psalm --show-info=false --diff
Scanning files...
Analyzing files...

░░

ERROR: InvalidReturnType - common/Yii.php:14:28 - The declared return type 'int' for Yii::foo is incorrect, got 'string'
    public function foo(): int {

ERROR: InvalidReturnStatement - common/Yii.php:15:16 - The type 'string(bar)' does not match the declared return type 'int' for Yii::foo
        return 'bar';

(OTHER ISSUES FROM THIS FILE, IGNORED THROUGH BASELINE FILE)

------------------------------
6 errors found
------------------------------

Checks took 3.45 seconds and used 276.309MB of memory
Psalm was able to infer types for 90.9566% of the codebase
```

This seems like a bug. All errors should be displayed only if `--ignore-baseline` option is provided.

**Comment by muglug** [comment]
Can antone reproduce on a public project?

**Comment by localheinz** [comment]
@muglug 

Yes, I can reproduce as well - take a look at https://github.com/ergebnis/factory-bot/pull/106:

- running an analysis locally succeeds
- running an analysis on GitHub Actions fails

When rebasing the branch locally, the first run fails, the second run succeeds.

Not sure what is going on.

**Comment by localheinz** [comment]
@muglug

Here's a reproducer!

Run

```shell
$ git clone git@github.com:ergebnis/factory-bot.git
$ cd factory-bot
$ make static-code-analysis
```

This works fine.

Now run:

```
$ git checkout fix/rename
$ make static-code-analysis
```

This fails with

```
vendor/bin/psalm --config=psalm.xml --diff --diff-methods --show-info=false --stats --threads=4
Scanning files...
Analyzing files...

░░░░░░░░░░░░░░░

ERROR: UndefinedClass - test/Unit/FixtureFactoryTest.php:220:21 - Class Ergebnis\FactoryBot\Test\DataProvider\NumberProvider does not exist (see https://psalm.dev/019)
    public function testGetListThrowsInvalidCountExceptionWhenCountIsLessThanOne(int $count): void


ERROR: UndefinedClass - test/Unit/FixtureFactoryTest.php:438:21 - Class Ergebnis\FactoryBot\Test\DataProvider\NumberProvider does not exist (see https://psalm.dev/019)
    public function testReferencedObjectsShouldBeCreatedAutomatically(int $count): void


ERROR: UndefinedClass - test/Unit/FixtureFactoryTest.php:467:21 - Class Ergebnis\FactoryBot\Test\DataProvider\NumberProvider does not exist (see https://psalm.dev/019)
    public function testReferencedObjectsShouldBeOverrideable(int $count): void


ERROR: UndefinedClass - test/Unit/FixtureFactoryTest.php:614:21 - Class Ergebnis\FactoryBot\Test\DataProvider\NumberProvider does not exist (see https://psalm.dev/019)
    public function testReferencedObjectsCanBeSingletons(int $count): void


------------------------------
4 errors found
------------------------------
35 other issues found.
You can display them with --show-info=true
------------------------------

Checks took 2.04 seconds and used 397.623MB of memory
Psalm was able to infer types for 99.6146% of the codebase
-----------------
100% test/Unit/FixtureFactoryTest.php (2 mixed)
100% test/Integration/FixtureFactoryTest.php (0 mixed)
100% src/FieldDefinition.php (0 mixed)
97% src/FixtureFactory.php (2 mixed)
100% src/Exception/EntityDefinitionNotRegistered.php (0 mixed)
100% test/Unit/Exception/EntityDefinitionNotRegisteredTest.php (0 mixed)
```

Then run

```
$ make static-code-analysis
```

and it passes with

```
vendor/bin/psalm --config=psalm.xml --diff --diff-methods --show-info=false --stats --threads=4
Scanning files...
Analyzing files...

░

------------------------------
No errors found!
------------------------------
21 other issues found.
You can display them with --show-info=true
------------------------------

Checks took 1.53 seconds and used 316.063MB of memory
Psalm was able to infer types for 99.6146% of the codebase
-----------------
97% src/FixtureFactory.php (2 mixed)
```

Note, it does not appear to be a cache issue - when I run

```
$ rm -rf .build
$ make static-code-analysis
```

it passes with

```
vendor/bin/psalm --config=psalm.xml --diff --diff-methods --show-info=false --stats --threads=4
Scanning files...
Analyzing files...

░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░

------------------------------
No errors found!
------------------------------
41 other issues found.
You can display them with --show-info=true
------------------------------
Psalm can automatically fix 1 of these issues.
Run Psalm again with
--alter --issues=MissingParamType --dry-run
to see what it can fix.
------------------------------

Checks took 5.35 seconds and used 255.823MB of memory
Psalm was able to infer types for 98.7848% of the codebase
-----------------
100% src/EntityDefinition.php (0 mixed)
100% src/Definition/Definitions.php (0 mixed)
100% src/Definition/AbstractDefinition.php (0 mixed)
81% src/FieldDefinition.php (5 mixed)
95% src/FixtureFactory.php (9 mixed)
100% src/Exception/InvalidCount.php (0 mixed)
100% src/Exception/InvalidDirectory.php (0 mixed)
100% src/Exception/InvalidFieldNames.php (0 mixed)
100% src/Exception/EntityDefinitionAlreadyRegistered.php (0 mixed)
100% src/Exception/InvalidDefinition.php (0 mixed)
100% src/Exception/EntityDefinitionNotRegistered.php (0 mixed)
100% test/Unit/FixtureFactoryTest.php (2 mixed)
100% test/Unit/EntityDefinitionTest.php (0 mixed)
100% test/Unit/Definition/FakerAwareDefinitionTest.php (0 mixed)
99% test/Unit/Definition/DefinitionsTest.php (1 mixed)
100% test/Unit/Definition/AbstractDefinitionTest.php (0 mixed)
100% test/Unit/FieldDefinitionTest.php (0 mixed)
100% test/Unit/AbstractTestCase.php (0 mixed)
100% test/Unit/Exception/InvalidDefinitionTest.php (0 mixed)
100% test/Unit/Exception/InvalidFieldNamesTest.php (0 mixed)
100% test/Unit/Exception/EntityDefinitionAlreadyRegisteredTest.php (0 mixed)
100% test/Unit/Exception/InvalidCountTest.php (0 mixed)
100% test/Unit/Exception/InvalidDirectoryTest.php (0 mixed)
100% test/Unit/Exception/EntityDefinitionNotRegisteredTest.php (0 mixed)
100% test/Util/Doctrine/ORM/EntityManagerFactory.php (0 mixed)
100% test/Fixture/FixtureFactory/Entity/User.php (0 mixed)
100% test/Fixture/FixtureFactory/Entity/Avatar.php (0 mixed)
100% test/Fixture/FixtureFactory/Entity/Project.php (0 mixed)
100% test/Fixture/FixtureFactory/Entity/Organization.php (0 mixed)
100% test/Fixture/FixtureFactory/Entity/Repository.php (0 mixed)
100% test/Fixture/Definition/Definitions/ExtendsAbstractDefinition/RepositoryDefinition.php (0 mixed)
100% test/Fixture/Definition/Definitions/IsAbstract/RepositoryDefinition.php (0 mixed)
100% test/Fixture/Definition/Definitions/Acceptable/RepositoryDefinition.php (0 mixed)
100% test/Fixture/Definition/Definitions/FakerAware/RepositoryDefinition.php (0 mixed)
100% test/Fixture/Definition/Definitions/FakerAware/OrganizationDefinition.php (0 mixed)
100% test/Fixture/Definition/Definitions/PrivateConstructor/RepositoryDefinition.php (0 mixed)
100% test/Fixture/Definition/Definitions/DoesNotImplementInterface/RepositoryDefinition.php (0 mixed)
100% test/Fixture/Definition/Definitions/ThrowsExceptionDuringConstruction/RepositoryDefinition.php (0 mixed)
100% test/Integration/FixtureFactoryTest.php (0 mixed)
100% test/Integration/AbstractTestCase.php (0 mixed)
100% test/DataProvider/NumberProvider.php (0 mixed)
100% test/AutoReview/SrcCodeTest.php (0 mixed)
```

When I then run

```
$ git checkout master
$ make static-code-analysis
```

it fails with

```
vendor/bin/psalm --config=psalm.xml --diff --diff-methods --show-info=false --stats --threads=4
Scanning files...
Analyzing files...

░░░░░░░░░░░░░░░

ERROR: UndefinedClass - test/Unit/FixtureFactoryTest.php:220:21 - Class Ergebnis\FactoryBot\Test\DataProvider\NumberProvider does not exist (see https://psalm.dev/019)
    public function testGetListThrowsInvalidCountExceptionWhenCountIsLessThanOne(int $count): void


ERROR: UndefinedClass - test/Unit/FixtureFactoryTest.php:438:21 - Class Ergebnis\FactoryBot\Test\DataProvider\NumberProvider does not exist (see https://psalm.dev/019)
    public function testReferencedObjectsShouldBeCreatedAutomatically(int $count): void


ERROR: UndefinedClass - test/Unit/FixtureFactoryTest.php:467:21 - Class Ergebnis\FactoryBot\Test\DataProvider\NumberProvider does not exist (see https://psalm.dev/019)
    public function testReferencedObjectsShouldBeOverrideable(int $count): void


ERROR: UndefinedClass - test/Unit/FixtureFactoryTest.php:614:21 - Class Ergebnis\FactoryBot\Test\DataProvider\NumberProvider does not exist (see https://psalm.dev/019)
    public function testReferencedObjectsCanBeSingletons(int $count): void


------------------------------
4 errors found
------------------------------
35 other issues found.
You can display them with --show-info=true
------------------------------

Checks took 2.08 seconds and used 397.622MB of memory
Psalm was able to infer types for 99.6146% of the codebase
-----------------
100% test/Unit/FixtureFactoryTest.php (2 mixed)
100% test/Integration/FixtureFactoryTest.php (0 mixed)
100% src/FieldDefinition.php (0 mixed)
97% src/FixtureFactory.php (2 mixed)
100% src/Exception/EntityDefinitionUnavailable.php (0 mixed)
100% test/Unit/Exception/EntityDefinitionUnavailableTest.php (0 mixed)
```

❗️ Looks like it can be reproduced by switching between branches that have different commits lined up.

**Comment by localheinz** [comment]
@muglug 

I'll leave the pull request unmerged for now, so you can take look.

Related, I have tried opening an issue because I think the error in regard to the data provider is a false positive.

**Comment by muglug** [comment]
Oh turns out this is fixed in latest Psalm master (which has a bunch of improvements to `--diff`/`--diff-method`).

**Comment by muglug** [comment]
Actually no it's not, completely.

**Comment by muglug** [comment]
Firstly, feel free to merge the PR – I can reproduce this trivially on master by clearing cache and just running Psalm on the single file `test/Unit/FixtureFactoryTest.php`.

---

This specific issue comes down to a hiccup around scanning, analysis and caching – it's not related to `--diff` explicitly.

The bug originates in the PHPUnit plugin, and it happens because the plugin doesn't provide sufficient information to Psalm when Psalm is in the process of performing static reflection on the codebase.

The `Psalm\Codebase::classOrInterfaceExists` check assumes Psalm has visited every possible file containing a class or interface.

When analysing the full project, this is true, but when you're just analysing that single file, Psalm doesn't know it also needs to scan the file that defines `Ergebnis\FactoryBot\Test\DataProvider\NumberProvider`.

The plugin should do something like this: https://github.com/vimeo/psalm/blob/93735712f0f101ad4ea939fa9d915aba19c7b059/src/Psalm/Internal/PhpVisitor/ReflectorVisitor.php#L2093-L2094

Those two lines tell Psalm when it's scanning a file (essentially performing static reflection) that the file references a classlike that should also be scanned, if Psalm is to understand the file's contents properly.

That doesn't currently happen for `@dataProvider` annotations. For this to work as expected, The plugin would need to have some extra code in `afterClassLikeVisit` that parses each method's docblock and ensures that Psalm knows it needs to also scan classes referenced in `@dataProvider`.

cc @weirdan

**Comment by weirdan** [comment]
@muglug so, to reproduce this, I would need:
1. Reference external data provider
2. Run psalm on the individual test file

right?

**Comment by muglug** [comment]
Yes!

**Comment by muglug** [comment]
Sorry, fat fingers

**Comment by localheinz** [comment]
Thank you, @muglug and @weirdan, for taking a look at this!

**Comment by muglug** [comment]
This is fixed I think!
